### ISSUE TITLE
Autowiring issue

### ISSUE THREAD

**Issue by jenssegers** [thephpleague/container#78]
I'm having some issues with the `ReflectionContainer` and autowiring. I tried digging into the code to do a PR, but I could not really figure out the problem *yet*.

In my example, I share a specific instance of a class first:

```php
$container = new League\Container\Container;

$container->delegate(
    new League\Container\ReflectionContainer
);

$container->share('App\Client', function()
{
    return new App\Client('test', 'test');
});
```

And I have another class called `App\SDK` which requires a `App\Client` instance from the constructor. Here comes the bad part. When I try to create an instance of `App\SDK` using reflection, I'm getting a lot of errors:

```php
$container->get('App\SDK');

PHP Fatal error:  Uncaught exception 'League\Container\Exception\NotFoundException' with message 'Unable to resolve a value for parameter (username) in the function/method (__construct)' in .../vendor/league/container/src/Argument/ArgumentResolverTrait.php:66
```

It seems like it tries to create the `App\Client` even though I have already bound an instance of it to the container. While digging into the code, I have noticed that `$this->getContainer()` returns `NULL` in the `ArgumentResolverTrait`. Shouldn't this contain the original container instance?

**Comment by localheinz** [comment]
@jenssegers 

Can you provide a test case?

**Comment by jenssegers** [comment]
I'll try to add it to the unit tests.

**Comment by jenssegers** [comment]
I was able to fix the issue by adding this to the `delegate` method from the Container:

```
    /**
     * Delegate a backup container to be checked for services if it
     * cannot be resolved via this container.
     *
     * @param  \Interop\Container\ContainerInterface $container
     * @return $this
     */
    public function delegate(InteropContainerInterface $container)
    {
        if ($container instanceof ImmutableContainerAwareInterface || $container instanceof ContainerAwareInterface) {
            $container->setContainer($this);
        }

        $this->delegates[] = $container;

        return $this;
    }
```

But it doesn't feel pretty...

**Comment by jenssegers** [comment]
Is there a reason why there are 2 almost identical interfaces `ContainerAwareInterface` and `ImmutableContainerAwareInterface`?

**Comment by philipobenito** [comment]
Yes, because the containers they accept are different, I think I know the issue but not at a computer so will fix later



—

**Comment by jenssegers** [comment]
Unit test as example:

```
    public function testDelegateSharesContainer()
    {
        $container = new Container;

        $container->delegate(new \League\Container\ReflectionContainer);

        $container->share('League\Container\Test\Asset\Bar', function () {
            $bar = new Asset\Bar();
            $bar->setSomething('bar');
            return $bar;
        });

        $bar = $container->get('League\Container\Test\Asset\Bar');
        $foo = $container->get('League\Container\Test\Asset\Foo');
        $this->assertEquals($foo->bar, $bar);
    }
```

**Comment by philipobenito** [comment]
Thanks, will get to it asap 



—

**Comment by localheinz** [comment]
@jenssegers 

Your solution is the way, though - see https://github.com/thephpleague/container/pull/79!

**Comment by jenssegers** [comment]
Thanks guys! Looking forward to the next tag! :)

**Comment by philipobenito** [comment]
Pushed and tagged along with another fix cc8ff4811351d7c94a01fe3d1f720111f98c8f66
